File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
def get_letters(num):
    letters = {
        2: ["a", "b", "c"],
        3: ["d", "e", "f"],
        4: ["g", "h", "i"],
        5: ["j", "k", "l"],
        6: ["m", "n", "o"],
        7: ["p", "q", "r", "s"],
        8: ["t", "u", "v"],
        9: ["w", "x", "y", "z"],
        
    }

    return letters.get(num, [])

class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        numbers=[int(digit) for digit in digits]
        lists=[]
        for number in numbers:
            lists.append(get_letters(number))  

        result=[""]

        for List in lists:
            new_result=[]
            for existing in result:
                for char in List:
                    new_result.append(existing + char)
            
            result=new_result
        
        return result


        

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *appendChar(char *existing, char ch)
{
    int len = strlen(existing);

    char *newString = malloc(len + 2);

    strcpy(newString, existing);

    newString[len] = ch;
    newString[len + 1] = '\0';

    return newString;
}


void processGroup(
    char **result,
    int resultSize,
    char *group,
    char ***newResult,
    int *newSize
)
{
    *newSize = 0;
    *newResult = NULL;

    for (int i = 0; i < resultSize; i++)
    {
        for (int j = 0; group[j] != '\0'; j++)
        {
            (*newSize)++;

            *newResult = realloc(
                *newResult,
                (*newSize) * sizeof(char *)
            );

            (*newResult)[*newSize - 1] =
                appendChar(result[i], group[j]);
        }
    }
}


char** letterCombinations(char* digits, int* returnSize)
{
    char *mapping[] = {
        "",
        "",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    };

    *returnSize = 0;

    if (digits == NULL || digits[0] == '\0')
        return NULL;

    char **result = malloc(sizeof(char *));
    int resultSize = 1;

    result[0] = malloc(1);
    result[0][0] = '\0';

    for (int i = 0; digits[i] != '\0'; i++)
    {
        char *group = mapping[digits[i] - '0'];

        char **newResult;
        int newSize;

        processGroup(
            result,
            resultSize,
            group,
            &newResult,
            &newSize
        );

        for (int j = 0; j < resultSize; j++)
        {
            free(result[j]);
        }

        free(result);

        result = newResult;
        resultSize = newSize;
    }

    *returnSize = resultSize;

    return result;
}
